catalyst/contrib/nn/criterion/supervised_contrastive.py

```diff
@@ -0,0 +1,88 @@
+from math import e
+
+import torch
+from torch import nn
+
+
+class SupervisedContrastiveLoss(nn.Module):
+    """A Contrastive embedding loss that uses targets.
+
+    It has been proposed in `Supervised Contrastive Learning`_.
+
+    .. _`Supervised Contrastive Learning`:
+        https://arxiv.org/pdf/2004.11362.pdf
+    """
+
+    def __init__(self, tau: float, reduction: str = "mean", pos_aggregation="in") -> None:
+        """
+        Args:
+            tau: temperature
+            reduction (string, optional): specifies the reduction to apply to the output:
+                ``"none"`` | ``"mean"`` | ``"sum"``.
+                ``"none"``: no reduction will be applied,
+                ``"mean"``: the sum of the output will be divided by the number of
+                positive pairs in the output,
+                ``"sum"``: the output will be summed.
+            pos_aggregation (string, optional): specifies the place of positive pairs aggregation:
+                ``"in"`` | ``"out"``.
+                ``"in"``: maximization of log(average positive exponentiate similarity)
+                ``"out"``: maximization of average positive similarity
+        Raises:
+            ValueError: if reduction is not mean, sum or none
+        """
+        super().__init__()
+        self.tau = tau
+        self.reduction = reduction
+        self.pos_aggregation = pos_aggregation
+
+        if self.reduction not in ["none", "mean", "sum"]:
+            raise ValueError(f"Reduction should be: mean, sum, none. But got - {self.reduction}!")
+        if self.pos_aggregation not in ["in", "out"]:
+            raise ValueError(
+                f"Positive aggregation should be: in or out. But got - {self.pos_aggregation}!"
+            )
+
+    def forward(self, features: torch.Tensor, targets: torch.Tensor) -> torch.Tensor:
+        """
+        Args:
+            features: [bs; feature_len]
+            targets: [bs]
+
+        Returns:
+            computed loss
+        """
+        # torch.exp(1/tau) self similarity
+        self_similarity = 1 / self.tau
+        exp_self_similarity = e ** (1 / self.tau)
```

Comment:
could we move them under `__init__`? 👀

Reply:
Do you think it would be better?
I don't have any preferences.

Reply:
it should be a bit-bit faster
or at least it should remove batch-independent code from forward pass

Reply:
hmm...
You are right.

Reply:
[Done](https://github.com/catalyst-team/catalyst/pull/1293/commits/792ec6d7d074246423f41678288a846c18865604)